tck/src/main/java/org/eclipse/microprofile/lra/tck/TckLRATypeTests.java

```diff
@@ -616,11 +616,11 @@ private String resourceRequest(String rootPath, String path, boolean startLRA, i
                 // validate that the method ran with an LRA and that it is still active
                 assertNotEquals(testName.getMethodName() + ": Resource method should not have run with an LRA: " + methodLraId,
                         0, methodLraId.length());
-                assertFalse(lraClient.isLRAFinished(URI.create(methodLraId)));
+                assertFalse(lraManagementService.isLRAFinished(URI.create(methodLraId)));
```

Comment:
No, because its SPI and its loaded through ServiceLoader.

Reply:
The SPI new method is defined in an interface:

```java
public interface LRAManagementService {
    boolean isLRAFinished(URI lraId);
}
```

and therefore the impact on the TCK will be minimal if the existing client operations class implements it.
I said in my first comment on this PR that `We should provide a default implementation in the TCK` and I gave the rationale of why.

Also, LRAClientOps contains more than this `isLRAFinished` method so it makes sense for LRAClientOps to use the new SPI method to implement the isLRAFinished method. This way the change to the TCK tests and resources will be transparent to client code.

If you take this approach (of changing how the methods in LRAClientOps are implemented) then if we need to replace other LRAClientOps methods with SPIs then the impact on the rest of the TCK code base will be minimal.

Reply:
I personally don't like the current implementation of the `isLRAFinished` method as it is. When I was developing some new test (#305) or cleaning some tests (#314) I have issues to understand what's happening (suddenly there are enlisted arbitrary participants which shows warning messages in log). As well the implementation is not bullet proof as it shows the LRA as finished when it was has not been started yet.

I beg for pardon but I don't remember what was the discussion on gitter (per #298) so maybe there was some details that I'm missing.
By me forcing implementation to implement its own isLRAFinished method is not an issue. While I think that it would be nice passing as less as possible burden to implementations.
I'm not sure if that's completely possible but what to use the boolean `isLRAFinished(URI lra, LRAMetricService lraMetricService, String resourceName)` instead of the simple `isLRAFinished(URI lra)` at all places?

Reply:
The issue isn't about whether the current implementation is good or bad, it is about providing the new SPI and still keeping the default implementation in the TCK. The reason for doing this is:

>  ... Then each implementation will still work and can implement this new SPI at their own leisure. I'm aware of at least one implementation which is on hold at the moment so it would be nice not to break their implementation by requiring them to implement this new SPI.

Forcing extra implementation requirements on these existing implementations (when we do not have to) is not  going to encourage them to return to the project.

Reply:
I understand the point but for me is connected. If the current implementation is incorrect then when the issues focus on the wrong method then the functionality should be addressed as well.

Nevertheless, what about the idea of using `isLRAFinished(URI lra, LRAMetricService lraMetricService, String resourceName)` instead of the `isLRAFinished(URI lra)` in tests?

Reply:
If are not to force the SPI then I agree with @ochaloup and we should replace everything with the metrics call. However, that will force each test resource to provide AfterLRA method and increment metrics which is another drawback.

I personally don't think that this new SPI would be an issue for any new or existent implementation as the other SPI we already have (LRARecoveryService) is way more complex in my opinion.

So to sum up options:
1/ force the implementation to implement new SPI
2/ provide SPI with available try-enlist code way to find out whether LRA is finished
3/ replace the calls with isLRAFinished based on metrics and add AfterLRA processors to each resource in TCK

From my perspective, the easiest 1/, cleanest 3/. So I vote for 1/ if we are to finish the spec sooner or 3/ if we some time to implement it.

Reply:
If it's an option I would like not adding new requirement to TCK implementers while fixing the current wrong behaviour.
I would vote for 3/.

Reply:
There is a fourth option:

4. Keep the status quo.

Reply:
But if it is a vote and we think option 3 is cleaner than the status quo and if someone is prepared to implement it then I agree with Ondra that we should choose option 3.